### DISCUSSION TITLE
Update homekit based on real button press

### DISCUSSION THREAD

**Discussion by bramboe** [HomeSpan/HomeSpan#382] [Q&A]
Hey!

So far I've been able to control a physical device by tapping in the physical pushbuttons. The only problem I seem to have right now is if the device itself is being controlled - pushing the actual buttons - and update Homekit. I tried to use `void loop` and look at when the `input` would change and than update a `int` to` true -or- false` and use that in my `boolean update`. This doesn't seem to work. 

Any idea how I could approach this?

**Comment by HomeSpan** [comment]
See Example 15.  It shows how to use the setVal() method to update HomeKit based on button presses.

**Reply by bramboe** [reply]
I assume you're referring to `SpanButton`? Does SpanButton take care of the `digitalRead`?

**Reply by HomeSpan** [reply]
Yes, it's all in there, though you don't have to use SpanButton if you already have your own code that reads button presses and changes the state of different variables.   The point of setVal(), whether used in SpanButton or anywhere else, is to communicate changes made on the device back to the Home App.

**Reply by bramboe** [reply]
I was using this:

```
  void powerOn() { // same for powerOff() but different value
    if (power->getVal() == 0) {
      power->setVal(1);
      digitalWrite(this->powerPin, HIGH);
      delay(250);    
      digitalWrite(this->powerPin, LOW);
      delay(1000);   
      digitalWrite(this->powerPin, HIGH);
      delay(250);         
      LOG1("Requested switch ON the device.\n");  
    }
  }  
```

```
  void loop() {
   
   int buttonState;

      if(digitalRead(powerLed) == LOW && power->getVal() == 0) {
      powerOn();
      buttonState = 1;
      } else if(digitalRead(powerLed) == LOW && power->getVal() == 1) {
      powerOff();
      buttonState = 0;
      }

etc..
```

But I think there is something wrong with the rest of my code as it seems to jump back to a previous state.

**Reply by HomeSpan** [reply]
Note that physical buttons "bounce" when pressed. This means they often switch on/off rapidly as the electrical contact is being made, and then again when it is released.  To address this you have to de-bounce buttons and switches by either adding some hardware (resistor/capacitor) or by filtering this noise in software.  

SpanButton() does this automatically for you in software.  Recommend you copy Example 15 and then modify to fit your specific device.

**Reply by bramboe** [reply]
Ok, looked again at my setup and improved it with the right resistor. Does `SpanButton` also take care of the `digitalWrite` by any chance?
